kafka-client/src/main/java/dev/responsive/kafka/internal/metrics/OrderedTagsSupplier.java

```diff
@@ -0,0 +1,127 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.metrics;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple wrapper around the tag map, so we can control the insertion order
+ * and make sure the tag ordering stays consistent with the mbean name
+ */
+public class OrderedTagsSupplier {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OrderedTagsSupplier.class);
+
+  // Base tags that all Responsive metrics are tagged with:
+  public static final String RESPONSIVE_VERSION_TAG = "responsive-version";
+  public static final String RESPONSIVE_COMMIT_ID_TAG = "responsive-commit-id";
+  public static final String STREAMS_VERSION_TAG = "streams-version";
+  public static final String STREAMS_COMMIT_ID_TAG = "streams-commit-id";
+  public static final String CONSUMER_GROUP_TAG = "consumer-group";
+  public static final String STREAMS_APPLICATION_ID_TAG = "streams-application-id";
+  public static final String STREAMS_CLIENT_ID_TAG = "streams-client-id";
+
+  // Group tags that are specific to the given metric group and scope
+  public static final String THREAD_ID_TAG = "thread-id";
+  public static final String TOPIC_TAG = "topic";
+  public static final String PARTITION_TAG = "partition";
+  public static final String STORE_NAME_TAG = "store-name";
+
+  private final String responsiveClientVersion;
+  private final String responsiveClientCommitId;
+  private final String streamsClientVersion;
+  private final String streamsClientCommitId;
+  private final String consumerGroup;
+  private final String streamsApplicationId;
+  private final String streamsClientId;
+  private final List<Entry<String, ?>> orderedUserTags;
+
+  public OrderedTagsSupplier(
+      final String responsiveClientVersion,
+      final String responsiveClientCommitId,
+      final String streamsClientVersion,
+      final String streamsClientCommitId,
+      final String consumerGroup,
+      final String streamsApplicationId,
+      final String streamsClientId,
+      final Map<String, ?> userTags
+  ) {
+    this.responsiveClientVersion = responsiveClientVersion;
+    this.responsiveClientCommitId = responsiveClientCommitId;
+    this.streamsClientVersion = streamsClientVersion;
+    this.streamsClientCommitId = streamsClientCommitId;
+    this.consumerGroup = consumerGroup;
+    this.streamsApplicationId = streamsApplicationId;
+    this.streamsClientId = streamsClientId;
+    this.orderedUserTags = userTags.entrySet().stream().sorted().collect(Collectors.toList());
+  }
+
+  public LinkedHashMap<String, String> applicationGroupTags() {
```

Comment:
we may want to consider an `EnumMap` here and have all the tags be enums with string values - `EnumMaps` are always maintained in natural order so it would be a global ordering that's all maintained in one place (the enum).

It also makes testing it pretty easy because each enum can be a name + an int for its ordinal (e.g. `FOO("foo", 1)` and you just test that `foo.ordinal() == foo.order()`

Reply:
Oh nice, I hadn't heard of EnumMap before

Great suggestion, but if you don't mind, I think I'll merge this for now so I can focus on the other release tasks -- I'll follow up on this next week

Reply:
👍 sounds good